File: Programmers/Binary_Search/입국심사.cpp
#include <vector>
#include <algorithm>
#define ll long long
using namespace std;

ll solution(int n, vector<int> v) {
    sort(v.begin(), v.end());
    ll ans = 0,
        left = 1,
        right = v[v.size()-1] * (ll) n;
    
    while(left <= right) {
        ll time = (left + right) / 2;
        ll sum = 0;
        
        for(int i = 0; i < v.size(); i++) 
            sum += time / v[i];

        if(sum >= n) {
            ans = time;
            right = time - 1;
        }
        else left = time + 1;
    }
    return ans;
}

File: Programmers/Binary_Search/입국심사.js
function solution(n, times) {
  const arr = times.sort();
  let ans = 0,
    l = 1,
    r = arr[times.length - 1] * n;

  while (l <= r) {
    let m = Math.floor((l + r) / 2);
    let sum = 0;

    for (let i of arr) {
      sum += Math.floor(m / i);
    }

    if (sum >= n) {
      r = m - 1;
      ans = m;
    } else {
      l = m + 1;
    }
  }
  return ans;
}
